cyten/models/degrees_of_freedom.py

```diff
@@ -0,0 +1,759 @@
+        Nmax = np.asarray(Nmax, dtype=int)
+        assert np.min(Nmax) > 0, (f'Invalid Nmax: {Nmax}; each boson species must have a max. '
+                                  'occupation number of at least 1')
+        self.Nmax = Nmax
+
+        DegreeOfFreedom.__init__(
+            self, leg=leg, state_labels=state_labels, onsite_operators=onsite_operators,
+            backend=backend, default_device=default_device
+        )
+
+    def test_sanity(self):
+        super().test_sanity()
+        for i in range(self.num_species):
+            N_i = self.creators[:, :, i] @ self.annihilators[:, :, i]
+            # check commutation relations
+            # BBd is 0 when going over the maximum occupation -> set this manually here
+            BBd = self.annihilators[:, :, i] @ self.creators[:, :, i]
+            mask = np.isclose(np.diag(BBd), 0)
+            BBd[mask, mask] += self.Nmax[i] + 1
+            assert np.allclose(BBd - N_i, np.eye(self.leg.dim))
+            # N_i has integer eigenvalues and is diagonal
+            N_i_rounded = np.around(N_i, 0)
+            assert np.allclose(N_i_rounded, N_i)
+            assert np.allclose(np.diag(np.diag(N_i)), N_i)
+            assert np.min(N_i_rounded) == 0
+            assert np.max(N_i_rounded) == self.Nmax[i]
+
+            # check commutation relations among different species
+            for j in range(i):
+                BiBdj = self.annihilators[:, :, i] @ self.creators[:, :, j]
+                BdjBi = self.creators[:, :, j] @ self.annihilators[:, :, i]
+                assert np.allclose(BiBdj, BdjBi)
+                BiBj = self.annihilators[:, :, i] @ self.annihilators[:, :, j]
+                BjBi = self.annihilators[:, :, j] @ self.annihilators[:, :, i]
+                assert np.allclose(BiBj, BjBi)
+                BdiBdj = self.creators[:, :, i] @ self.creators[:, :, j]
+                BdjBdi = self.creators[:, :, j] @ self.creators[:, :, i]
+                assert np.allclose(BdiBdj, BdjBdi)
+
+    def add_individual_occupation_ops(self, are_diagonal: bool = True):
+        """Add occupation and parity operators for each species as symmetric onsite operators.
+        
+        The added operators include::
+            - occupation operators for each boson species `Ni` for species `i`
+                / `N` for a single species
+            - parity operators for each boson species `Pi` for species `i`
+                / `P` for a single species
+            - squared occupation operators for each boson species `NiNi` for species `i`
+                / `NN` for a single species
+
+        Parameters
+        ----------
+        are_diagonal : bool
+            Whether or not the constructed operators are diagonal, such that they can be
+            represented as :class:`DiagonalTensor`.
+        """
+        for i in range(self.num_species):
+            op_names = ['N', 'NN', 'P'] if self.num_species == 1 else [f'N{i}', f'N{i}N{i}', f'P{i}']
```

Comment:
can we improve this case distinction? Maybe always have ``N0`` and add ``N`` as an alias if single-species?